Add MLCD model

tanhuajie opened the pull request:
# What does this PR do?

This PR adds **MLCD** model from DeepGlint-AI Team.

Fixes #36181


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@amyeroberts @qubvel @ArthurZucker 

## Quick Test

```python
from transformers import AutoProcessor, MLCDVisionModel
from PIL import Image
import requests

# Load model and processor
model = MLCDVisionModel.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")
processor = AutoProcessor.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")

# Process single image
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
inputs = processor(images=image, return_tensors="pt")

# Get visual features
outputs = model(**inputs)
features = outputs.last_hidden_state

print(f"Extracted features shape: {features.shape}")
```

tanhuajie committed:
Add MLCD model

tanhuajie committed:
Update codes for auto-mapping

tanhuajie committed:
Add test scripts for MLCD

tanhuajie committed:
Update doc for MLCD model

tanhuajie committed:
Fix import error

tanhuajie committed:
Fix import error

tanhuajie committed:
Fix CI error for attention_outputs

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix CI error for initialization

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Fix code style for CI

tanhuajie committed:
Reformat codes and docs for CI test

tanhuajie committed:
Reformat codes and docs for CI test

tanhuajie committed:
Remove unused attributes for CI test

tanhuajie committed:
Fix style for CI test

tanhuajie committed:
List MLCD in flash_attn doc

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
![image](https://github.com/user-attachments/assets/0a61c221-f0ef-4b2f-9d59-c8fc7eede9b1)

Hi, Pavel @qubvel. Hope this message finds you well. It appears that our PR failed on the final step during CI. From the error messages, it looks like the issue stems from the Rt-Detr tests rather than problems within our code. Could you please guide us on how to disable or skip the tests for other models so we can successfully complete the CI process for our PR? Thanks!

anxiangsir commented:
@qubvel @Rocketknight1 

Could you please take a moment to review the pull request? Your insights would be greatly appreciated and would help ensure the quality of the changes. We're truly grateful for your help!

tanhuajie committed:
Merge branch 'main' into add_mlcd

Rocketknight1 commented:
Gentle ping @qubvel, but let me know if you want me to take any part of the review!

qubvel commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+        Returns:
+        """
+
+        twh = (1, pixel_values.size(3) // self.config.patch_size, pixel_values.size(2) // self.config.patch_size)
```
please use more descriptive variable name, not `tbh`

qubvel commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+            last_hidden_state=last_hidden_state,
+            pooler_output=pooled_output,
+            hidden_states=encoder_outputs.hidden_states,
+            # attentions=encoder_outputs.attentions,
```
no commented code

qubvel commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+# coding=utf-8
+# Copyright 2024 Mistral and the HuggingFace Inc. team. All rights reserved.
```
Everywhere

```suggestion
# Copyright 2025 Mistral and the HuggingFace Inc. team. All rights reserved.
```

tanhuajie replied:
Fixed.

qubvel commented on tests/models/mlcd/test_modeling_mlcd.py:
```diff
@@ -0,0 +1,152 @@
+
+
+@require_torch
+class MLCDVisionModelModelTest(ModelTesterMixin, unittest.TestCase):
```
```suggestion
class MLCDVisionModelTest(ModelTesterMixin, unittest.TestCase):
```

tanhuajie replied:
Thanks. Fixed now.

qubvel commented on tests/models/mlcd/test_modeling_mlcd.py:
```diff
@@ -0,0 +1,152 @@
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```

tanhuajie replied:
Fixed.

qubvel commented on tests/models/mlcd/test_modeling_mlcd.py:
Please add Integration tests as well

tanhuajie replied:
Got it. Integration tests for `MLCD` have been added in `test_modeling_mlcd.py`

qubvel commented on src/transformers/models/mlcd/convert_mlcd_weights_to_hf.py:
```diff
@@ -0,0 +1,215 @@
+)
+from transformers.modeling_outputs import BaseModelOutputWithPooling
+
+from .original_vit_rope2d import RoPE2d_ViT_bigG_14_1024
+
+
+def copy_attn_layer(hf_attn_layer, pt_attn_layer):
+    # self.in_proj = nn.Linear(dim, dim * 3, bias=True)
+    # self.out_proj = nn.Linear(dim, dim)
+
+    q_proj, k_proj, v_proj = pt_attn_layer.in_proj.weight.data.chunk(3, dim=0)
+    q_proj_bias, k_proj_bias, v_proj_bias = pt_attn_layer.in_proj.bias.data.chunk(3, dim=0)
+
+    out_proj_weights = pt_attn_layer.out_proj.weight
+    out_proj_bias = pt_attn_layer.out_proj.bias
+
+    hf_attn_layer.q_proj.weight.data = q_proj
```
Please see other conversion scripts to structure your code (I would recommend `mllama`, because it follows out new standards for conversion). 

The general idea is that we do not use 3rd party model to convert weights, instead, we should 
1) load state dict
2) rename/split/transpose weights
3) load new state dict into HF model to check everything match

tanhuajie replied:
Thanks for reminder. `convert_mlcd_weights_to_hf.py` has been refactored with reference to both `mllama` and `siglip2`.

qubvel reviewed (commented):
Hey @tanhuajie! Sorry for the delay and thanks a lot for working on the model addition to `transformers`, great work, and already looks super clean!

I see the model is built on pretty standard modules, so it would be incredibly helpful if you could reuse library modules with inheritance using our new `modular` tool.

 - https://huggingface.co/docs/transformers/modular_transformers
 - see other models such as gemma/ijepa/siglip2 (`modular_*.py` file)

See other comments below!

tanaymeh commented:
> Hey @tanaymeh! Sorry for the delay and thanks a lot for working on the model addition to `transformers`, great work, and already looks super clean!
> 
> I see the model is built on pretty standard modules, so it would be incredibly helpful if you could reuse library modules with inheritance using our new `modular` tool.
> 
> * https://huggingface.co/docs/transformers/modular_transformers
> * see other models such as gemma/ijepa/siglip2 (`modular_*.py` file)
> 
> See other comments below!

I think you tagged the wrong person here @qubvel 😅

tanhuajie committed:
Fix: typos, modulars, refactors from suggestions

tanhuajie committed:
Refactoring convert_mlcd_weights_to_hf.py from suggestions

tanhuajie committed:
Fix: docs conflicts

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Fix error for CI test

tanhuajie committed:
Fix style for CI test

tanhuajie committed:
Add integration test for MLCD

tanhuajie commented on src/transformers/models/mlcd/original_vit_rope2d.py:
```diff
@@ -0,0 +1,383 @@
+# coding=utf-8
```
Yes, this file is redundant and has been removed.

tanhuajie commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+    sin = sin.unsqueeze(1).repeat(1, 1, 2).unsqueeze(0).float()
+    output = (tensor * cos) + (rotate_half(tensor) * sin)
+    output = output.to(orig_dtype)
+    return output
```
Thanks for this suggestion. The file `modeling_mlcd.py` has been refactored and regenerated using the `modular` tool (Nice Tool!!!). This function has been reused now.

tanhuajie commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+        attn_output = attn_output.view(seq_length, batch_size, -1)  # [seq_length, batch_size, embedding_dim]
+        attn_output = self.out_proj(attn_output)
+        attn_output = attn_output.permute(1, 0, 2).contiguous()  # [batch_size, seq_length, embedding_dim]
+        return attn_output, None
```
Yep, this pattern has also been reused now.

tanhuajie commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+        max_grid_size = grid_thw[:, 1:].max()
+        rotary_pos_emb_full = self.vision_rotary_embedding(max_grid_size)
+        rotary_pos_emb = rotary_pos_emb_full[pos_ids].flatten(1)
+        return rotary_pos_emb
```
1. You are right. I have already put it into `MLCDRotaryEmbedding` module.
2. Completed. Vectorize the process to avoid cycle !!

tanhuajie commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,520 @@
+    def forward(
+        self,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        # output_attentions: Optional[bool] = None,
```
Got it. This argument has been supported in the newest commit.

tanhuajie committed:
Merge branch 'main' into add_mlcd

qubvel reviewed (commented):
Hi @tanhuajie, thanks for addressing comments and using modular! I left a few more comments on how to use inheritance a bit more. Other than that looks pretty clean, great work 🤗

tanhuajie committed:
Refactoring by class inheritance

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+        key_states = key_states.permute(0, 2, 1, 3).contiguous()
+        value_states = value_states.permute(0, 2, 1, 3).contiguous()
+
+        attn_output = F.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=attention_mask,
+            dropout_p=self.dropout if self.training else 0.0,
+        )
+        attn_output = attn_output.permute(2, 0, 1, 3).contiguous()  # [seq_length, batch_size, num_heads, head_dim]
```
OK. Attention interface has been refactored in the latest commit.

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+        self.mlp = MLCDMLP(config)
+        self.layer_norm2 = nn.LayerNorm(self.embed_dim, eps=config.layer_norm_eps)
+
+    def forward(
```
Got it. But I notice this modification would cause the `config` parameter in `modeling_mlcd.py` to be automatically generated as `MLCDConfig` by the `modular` tool, rather than the desired `MLCDVisionConfig`. Therefore, I ultimately made the following changes instead:

```python
class MLCDEncoderLayer(CLIPEncoderLayer):
    def __init__(self, config: MLCDVisionConfig):
        super().__init__(config)
```

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
```
Completed !

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+        self.post_init()
+
+    def get_input_embeddings(self) -> nn.Module:
+        return self.vision_model.embeddings.patch_embedding
```
Completed.

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+        self.post_layernorm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
+
+        self.vision_rotary_embedding = MLCDRotaryEmbedding(config.hidden_size // config.num_attention_heads // 2)
+        self.class_pos_emb = nn.Parameter(torch.randn(1, config.hidden_size // config.num_attention_heads // 2))
```
Completed.

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,723 @@
+        )
+
+        self.num_patches = (self.image_size // self.patch_size) ** 2
+        self.num_positions = self.num_patches + 1
```
Completed.

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
Hi @qubvel! Thank you for your further suggestions. Based on your latest comments, I have made additional modifications to `modular_mlcd.py`. These changes include improving the reusability of the modular components through class inheritance and refactoring the Attention interface to support different attention implementations. 🤗

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
Hi @qubvel, just a gentle reminder to review the latest commits. Your feedback is incredibly valuable, and I’d appreciate it if you could take a look when you have a moment. Thanks! 🤗

qubvel commented on tests/models/mlcd/test_modeling_mlcd.py:
```diff
@@ -0,0 +1,217 @@
+    test_pruning = False
+    test_head_masking = False
+    test_torchscript = False
+    test_resize_embeddings = False
```
```suggestion
    test_resize_embeddings = False
    test_torch_exportable = True
```

tanhuajie replied:
`MLCD` is a vision-only model and doesn't have a `vocab_size` that needs to be adjusted, so I think, it's better to keep `test_resize_embeddings = False`.

qubvel replied:
Yes, I'm asking to add `test_torch_exportable=True`

qubvel reviewed (commented):
Thanks for addressing the comments! I think we are close to make it done, added a few more comments, please have a look!

tanhuajie committed:
Fix: refactor attention interface, adjust codes

tanhuajie committed:
Fix: merging conflicts

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Fix: merging conflicts

tanhuajie committed:
Fix: style for CI test

tanhuajie committed:
Fix: style for CI test

tanhuajie committed:
Fix: set test_resize_embeddings to be False

tanhuajie commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,762 @@
+    def forward(self, grid_thw: torch.Tensor) -> torch.Tensor:
+        """Calculate sequence length from grid, and then get the RoPE for MLCDVisionModel"""
+
+        t, h, w = grid_thw[:, 0], grid_thw[:, 1], grid_thw[:, 2]
```
Yes. I have adjusted them.

tanhuajie committed:
Fix: initializer for CI test

tanhuajie commented:
Hey @qubvel, Thanks for your comments! I have refactored the attention interface and adjusted the code based on your suggestions. I think it's ready for your review again. Please take a look when you have time! Thanks 🤗

anxiangsir commented:
Hi @qubvel ,

Thanks again for your valuable feedback earlier! We’ve refactored the attention interface and made adjustments to the code based on your suggestions. It should be ready for your review now. Whenever you have some time, we’d really appreciate it if you could take a look. Thank you so much!

anxiangsir commented:
Hi @qubvel, hope you're doing well! Just a quick follow-up on the updated interface based on your feedback—would love to hear your thoughts when you have time.

Thanks again!

qubvel commented:
Hi @tanhuajie, thanks for the ping, I will review it today or tomorrow 🤗

anxiangsir commented:
Hi @qubvel ，Just a gentle reminder—when you have a moment, please help us review the updated interface. Would love to hear your feedback!  🤗🤗🤗

qubvel commented on src/transformers/models/mlcd/modular_mlcd.py:
````diff
@@ -0,0 +1,599 @@
+    ) -> Union[Tuple, BaseModelOutputWithPooling]:
+        r"""
+        ```python
+        >>> from PIL import Image
+        >>> import requests
+        >>> from transformers import AutoProcessor, MLCDVisionModel
+        >>> model = MLCDVisionModel.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")
+        >>> processor = AutoProcessor.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> inputs = processor(images=image, return_tensors="pt")
+        >>> outputs = model(**inputs, output_attentions=True)
+        >>> features = outputs.last_hidden_state
+        >>> print(f"Extracted features shape: {features.shape}")
+        >>> print(f"Number of attention layers: {len(outputs.attentions)}")
+        >>> print(f"Attention shape: {outputs.attentions[0].shape}")
````
```suggestion
        >>> import requests
        >>> from PIL import Image
        >>> from transformers import AutoProcessor, MLCDVisionModel

        >>> model = MLCDVisionModel.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")
        >>> processor = AutoProcessor.from_pretrained("DeepGlint-AI/mlcd-vit-bigG-patch14-336")
        
        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
        >>> image = Image.open(requests.get(url, stream=True).raw)
        >>> inputs = processor(images=image, return_tensors="pt")
        
        >>> with torch.no_grads():
        ...     outputs = model(**inputs, output_attentions=True)
        
        >>> features = outputs.last_hidden_state
        >>> print(f"Extracted features shape: {features.shape}")
        >>> print(f"Number of attention layers: {len(outputs.attentions)}")
        >>> print(f"Attention shape: {outputs.attentions[0].shape}")
```

qubvel commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,599 @@
+        query_states = self.q_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        key_states = self.k_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        value_states = self.v_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        if position_embeddings is None:
+            logger.warning_once(
+                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
+                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
+                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
+                "removed and `position_embeddings` will be mandatory."
+            )
+            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
+            cos = emb.cos().unsqueeze(0).float()
+            sin = emb.sin().unsqueeze(0).float()
+        else:
+            cos, sin = position_embeddings
```
Which way is actually used? Not sure if we need a warning for a new model, can we avoid it?

qubvel replied:
@tanhuajie I mean, can we avoid `if position_embeddings is None` path completely?

qubvel reviewed (approved):
Nice! It seems ready, just a few nit on my side! Thanks for iterating on the model implementation 🚀 

P.S. please resolve conflicts, merge main and make sure related tests pass 🤗

qubvel commented:
run-slow: mlcd

qubvel commented:
Also integraion test is failing

```
FAILED tests/models/mlcd/test_modeling_mlcd.py::MLCDVisionModelIntegrationTest::test_inference - RuntimeError: Input type (torch.FloatTensor) and weight type (torch.cuda.FloatTensor) should be the same or input should be a MKLDNN tensor and weight is a dense tensor
```

please, make sure to take into account `torch_device`

tanhuajie committed:
Fix: conflicts, CI test, warning and refactoring

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Fix: merging conflicts

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Refactor

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
run-slow: mlcd

tanhuajie committed:
Merge branch 'main' into add_mlcd

qubvel commented:
run-slow: mlcd

tanhuajie committed:
Update docs

qubvel commented on docs/source/en/model_doc/mlcd.md:
```diff
@@ -48,21 +48,24 @@ Result:
 inputs = processor(images=image, return_tensors="pt")
 
+# Generate outputs
+with torch.no_grads():
```
```suggestion
with torch.no_grad():
```

tanhuajie replied:
ohh, sorry...

tanhuajie committed:
Fix mistakes

tanhuajie commented:
Hi @qubvel, I think it ready to merge now. 🤗 Thanks again for your review!

qubvel commented:
run-slow: mlcd

qubvel commented:
@tanhuajie please see above comment regarding positional_embeddings

qubvel commented:
Also there is failing test for multi-GPU setup, I suppose `_no_split_modules` attribute should be updated:

```
FAILED tests/models/mlcd/test_modeling_mlcd.py::MLCDVisionModelTest::test_model_parallelism - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1! (when checking argument for argument tensors in method wrapper_CUDA_cat)
```

Stacktrace:

```
self = MLCDVisionTransformer(
  (embeddings): MLCDVisionEmbeddings(
    (patch_embedding): Conv2d(3, 32, kernel_size=(2, 2), ...t_layernorm): LayerNorm((32,), eps=1e-05, elementwise_affine=True)
  (vision_rotary_embedding): MLCDRotaryEmbedding()
)
pixel_values = tensor([[[[7.4678e-01, 9.0169e-01, 5.9101e-01,  ..., 1.9737e-01,
           2.6010e-01, 3.8128e-01],
          [8.7451...         [2.5349e-01, 4.0613e-01, 3.7064e-01,  ..., 8.5400e-01,
           5.6[229](https://github.com/huggingface/transformers/actions/runs/14318436461/job/40129963836#step:12:230)e-01, 1.6790e-01]]]], device='cuda:0')
output_attentions = False, output_hidden_states = False, return_dict = True

    @add_start_docstrings_to_model_forward(MLCD_VISION_INPUTS_DOCSTRING)
    def forward(
        self,
        pixel_values: Optional[torch.FloatTensor] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
    ) -> Union[Tuple, BaseModelOutputWithPooling]:
        r"""
        Returns:
    
        """
        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
    
        if pixel_values is None:
            raise ValueError("You have to specify pixel_values")
    
        num_patches_height = pixel_values.shape[-2] // self.config.patch_size
        num_patches_width = pixel_values.shape[-1] // self.config.patch_size
        rotary_pos_emb = self.vision_rotary_embedding(num_patches_height, num_patches_width)
>       rotary_pos_emb = torch.cat([self.class_pos_emb, rotary_pos_emb], dim=0)
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1! (when checking argument for argument tensors in method wrapper_CUDA_cat)

src/transformers/models/mlcd/modeling_mlcd.py:536: RuntimeError
```

tanhuajie committed:
Remove unused args and fix multi-gpu error

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
> @tanhuajie please see above comment regarding positional_embeddings

Remove this unused args 'positional_embeddings' now 🤗

tanhuajie committed:
Merge branch 'main' into add_mlcd

qubvel commented:
run-slow: mlcd

qubvel commented:
Re the change, the idea is to use `position_embeddings` rather than `rotary_pos_emb` argument.

```python
        if position_embeddings is None:
            logger.warning_once(
                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
                "removed and `position_embeddings` will be mandatory."
            )
            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
            cos = emb.cos()
            sin = emb.sin()
        else:
            cos, sin = position_embeddings
```

So, 

```python
emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
position_embeddings = (emb.cos(), emb.sin())
```

Should happen not inside the attention, but inside the model itself and them propagated across layers

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Revert position_embeddings

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie commented:
> > Re the change, the idea is to use `position_embeddings` rather than `rotary_pos_emb` argument.
> > ```python
> >         if position_embeddings is None:
> >             logger.warning_once(
> >                 "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
> >                 "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
> >                 "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
> >                 "removed and `position_embeddings` will be mandatory."
> >             )
> >             emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
> >             cos = emb.cos()
> >             sin = emb.sin()
> >         else:
> >             cos, sin = position_embeddings
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So,
> > ```python
> > emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
> > position_embeddings = (emb.cos(), emb.sin())
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should happen not inside the attention, but inside the model itself and then have them propagated across layers
> 
> Okay, I'll make the revisions based on your suggestions later~

Hey @qubvel. I think it's done now. 🤗 Let me know if you have any feedback when you have a moment.

qubvel reviewed (approved):
Thanks for addressing comments 🤗

qubvel commented:
cc @ArthurZucker for review when you have bandwidth

anxiangsir commented:
Dear @ArthurZucker ,

I hope this message finds you well! I just wanted to kindly follow up and ask if you could help review this whenever you have the time. Please don’t hesitate to let me know if there’s anything I can provide to assist. I truly appreciate your time and support!

ArthurZucker commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,585 @@
+        self.hidden_act = hidden_act
+
+
+class MLCDConfig(MLCDVisionConfig):
+    pass
```
this is weird to me, we should not have two configs that are exactly the same

tanhuajie replied:
Hi, @ArthurZucker. The reason we provided a dummy `MLCDConfig` class is that when using the `modular` tool to create our custom `MLCDEncoder` by inheriting from `CLIPEncoder`, it automatically generates an unexpected `MLCDConfig` class. This generated configuration is identical to `MLCDVisualConfig` in all aspects except for the name. Therefore, we included the dummy `MLCDConfig` to handle this case.

If this approach isn't preferred, we can alternatively use the overwrite method to eliminate the dummy `MLCDConfig` and retain only `MLCDVisualConfig`. Here's how to implement it:

```python
class MLCDEncoder(CLIPEncoder):
    """
    Transformer encoder consisting of `config.num_hidden_layers` self attention layers. Each layer is a
    [`MLCDEncoderLayer`].
    Args:
        config: MLCDVisualConfig
    """

    def __init__(self, config: MLCDVisualConfig):
        super().__init__()
```

This modification explicitly specifies `MLCDVisualConfig` as the expected configuration type. 🤗

ArthurZucker commented on src/transformers/models/mlcd/modular_mlcd.py:
```diff
@@ -0,0 +1,585 @@
+        self.num_key_value_groups = config.num_key_value_groups
+        self.is_causal = False
+
+    def forward(
```
very nice!

ArthurZucker commented on src/transformers/models/mlcd/modular_mlcd.py:
Very nice! Kudos 🤗

ArthurZucker reviewed (approved):
Marvelous thanks all for the great PR! 🤗

tanhuajie commented:
> Hey, @tanhuajie let's resolve conflicts with main, and we are good to merge it!

Great, I'll resolve the conflicts right away. 🤗

tanhuajie committed:
Solve conflicts

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Solve conflicts

tanhuajie committed:
Remove dummy

tanhuajie committed:
Update _init_weights

tanhuajie committed:
Merge branch 'main' into add_mlcd

tanhuajie committed:
Update _init_weights

tanhuajie committed:
Update _init_weights for CI test

tanhuajie commented:
Hi @qubvel, I've resolved the conflicts, removed the dummy `MLCDConfig`, and refined `_init_weights()`. Then, I think MLCD should now be ready to merge. 🤗cc @anxiangsir @ArthurZucker

qubvel commented:
run-slow: mlcd

qubvel commented:
Thanks a lot for iterating on the model addition! Gerat work 🤗 
Merging 🚀

tanhuajie commented on src/transformers/models/mlcd/modeling_mlcd.py:
```diff
@@ -0,0 +1,679 @@
+    def get_input_embeddings(self) -> nn.Module:
+        return self.vision_model.embeddings.patch_embedding
+
+    @add_start_docstrings_to_model_forward(MLCD_VISION_INPUTS_DOCSTRING)
```
Thank you for the reminder. I'll take a look at how to resolve the issue you mentioned.

tanhuajie replied:
Thanks for a fix ❤️
